flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java

```diff
@@ -0,0 +1,353 @@
+          reuse.set(i, source.readUTF());
+          break;
+        case UUID:
+          long mostSignificantBits = source.readLong();
+          long leastSignificantBits = source.readLong();
+          reuse.set(i, new UUID(mostSignificantBits, leastSignificantBits));
+          break;
+        case FIXED:
+        case BINARY:
+          byte[] bytes = new byte[source.readInt()];
+          source.read(bytes);
+          reuse.set(i, ByteBuffer.wrap(bytes));
+          break;
+        case DECIMAL:
+          byte[] unscaledBytes = new byte[source.readInt()];
+          source.read(unscaledBytes);
+          int scale = source.readInt();
+          BigDecimal decimal = new BigDecimal(new BigInteger(unscaledBytes), scale);
+          reuse.set(i, decimal);
+          break;
+        case STRUCT:
+        case MAP:
+        case LIST:
+        default:
+          // SortKey transformation is a flattened struct without list and map
+          throw new UnsupportedOperationException(
+              String.format("Field %d has unsupported field type: %s", fieldId, typeId));
+      }
+    }
+
+    return reuse;
+  }
+
+  @Override
+  public void copy(DataInputView source, DataOutputView target) throws IOException {
+    // no optimization here
+    serialize(deserialize(source), target);
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (!(obj instanceof SortKeySerializer)) {
+      return false;
+    }
+
+    SortKeySerializer other = (SortKeySerializer) obj;
+    return Objects.equals(schema.asStruct(), other.schema.asStruct())
+        && Objects.equals(sortOrder, other.sortOrder);
+  }
+
+  @Override
+  public int hashCode() {
+    return schema.asStruct().hashCode() * 31 + sortOrder.hashCode();
+  }
+
+  @Override
+  public TypeSerializerSnapshot<SortKey> snapshotConfiguration() {
+    return new SortKeySerializerSnapshot(schema, sortOrder);
+  }
+
+  public static class SortKeySerializerSnapshot implements TypeSerializerSnapshot<SortKey> {
+    private static final int CURRENT_VERSION = 1;
+
+    private Schema schema;
+    private SortOrder sortOrder;
+
+    @SuppressWarnings({"checkstyle:RedundantModifier", "WeakerAccess"})
+    public SortKeySerializerSnapshot() {
+      // this constructor is used when restoring from a checkpoint.
+    }
+
+    // constructors need to public. Otherwise, Flink state restore would complain
+    // "The class has no (implicit) public nullary constructor".
+    @SuppressWarnings("checkstyle:RedundantModifier")
+    public SortKeySerializerSnapshot(Schema schema, SortOrder sortOrder) {
+      this.schema = schema;
+      this.sortOrder = sortOrder;
+    }
+
+    @Override
+    public int getCurrentVersion() {
+      return CURRENT_VERSION;
+    }
+
+    @Override
+    public void writeSnapshot(DataOutputView out) throws IOException {
+      Preconditions.checkState(schema != null, "Invalid schema: null");
+      Preconditions.checkState(sortOrder != null, "Invalid sort order: null");
```

Comment:
How about using `Preconditions.checkNotNull` to check if schema and sortOrder are null or not
Same for line 329, 330

Reply:
I personally prefer to avoid `chechNotNull`, as the result is a `NullPointerException`. Generally `NullPointerExceptions` are something very unexpected, while `IllegalArgumentExceptions` are a different category for me (configuration errors, and such)

That is why, I did not flag this at the review, but - taking a second look - this could be strange for someone less biased